File: VAMPObjCSample/VAMPObjCSample/InfoViewController.m
//
//  InfoViewController.m
//  VAMPObjCSample
//
//  Created by Supership Inc. on 2023/01/28.
//  Copyright © 2023 Supership Inc. All rights reserved.
//

#import <CoreTelephony/CTCarrier.h>
#import <CoreTelephony/CTTelephonyNetworkInfo.h>
#import <Foundation/Foundation.h>

#import <VAMP/VAMP.h>

#import "InfoViewController.h"

#define b2s(flag) (flag ? @"true" : @"false")

typedef NSString *(*AdNetworkVersionMethod)(id, SEL);
typedef NSString *(*AdapterVersionMethod)(id, SEL);

@interface InfoViewController ()

@property (nonatomic) UITextView *adInfoView;

@end

@implementation InfoViewController

+ (instancetype) instantiateViewController {
    return [InfoViewController new];
}

- (instancetype) init {
    if (self = [super init]) {
        _adInfoView = [UITextView new];
        _adInfoView.textContainerInset = UIEdgeInsetsMake(0, 16.f, 0, 16.f);
        _adInfoView.editable = NO;
        _adInfoView.selectable = NO;
    }

    return self;
}

- (void) viewDidLoad {
    [super viewDidLoad];

    self.title = @"Info";

    [self.view addSubview:self.adInfoView];

    self.adInfoView.text = @"";

    // VAMP SDK Version
    [self addInfoText:[NSString stringWithFormat:@"VAMPSDK: %@",
                       VAMPSDKVersion]];

    // アドネットワークバージョン
    [self addInfoText:[NSString stringWithFormat:@"AdMobSDK: %@",
                       [self adNetworkVersion:@"VAMPAdMobSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"maioSDK: %@",
                       [self adNetworkVersion:@"VAMPMaioSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"UnityAdsSDK: %@",
                       [self adNetworkVersion:@"VAMPUnityAdsSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"PangleSDK: %@",
                       [self adNetworkVersion:@"VAMPPangleSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"LINEAdsSDK: %@",
                       [self adNetworkVersion:@"VAMPLINEAdsSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"IronSourceSDK: %@",
                       [self adNetworkVersion:@"VAMPIronSourceSDKAdapter"]]];

    [self addInfoText:@"\n"];

    // アダプタバージョン
    [self addInfoText:[NSString stringWithFormat:@"AdMob Adapter: %@",
                       [self adapterVersion:@"VAMPAdMobSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"maio Adapter: %@",
                       [self adapterVersion:@"VAMPMaioSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"UnityAds Adapter: %@",
                       [self adapterVersion:@"VAMPUnityAdsSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"Pangle Adapter: %@",
                       [self adapterVersion:@"VAMPPangleSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"LINEAds Adapter: %@",
                       [self adapterVersion:@"VAMPLINEAdsSDKAdapter"]]];
    [self addInfoText:[NSString stringWithFormat:@"IronSource Adapter: %@",
                       [self adapterVersion:@"VAMPIronSourceSDKAdapter"]]];

    [self addInfoText:@"\n"];

    [self addInfoText:[NSString stringWithFormat:@"useHyperID: %@",
                       b2s(VAMP.useHyperID)]];

    [self addInfoText:@"\n"];

    // デバイス名
    [self addInfoText:[NSString stringWithFormat:@"device name: %@", [UIDevice currentDevice].name]];

    // OS名
    [self addInfoText:[NSString stringWithFormat:@"system name: %@", [UIDevice currentDevice].systemName]];

    // OSバージョン
    [self addInfoText:[NSString stringWithFormat:@"system version: %@", [UIDevice currentDevice].systemVersion]];

    // OSモデル
    [self addInfoText:[NSString stringWithFormat:@"OS model: %@", [UIDevice currentDevice].model]];

    // キャリア情報
    CTTelephonyNetworkInfo *networkInfo = [[CTTelephonyNetworkInfo alloc] init];
    CTCarrier *provider = [networkInfo subscriberCellularProvider];
    [self addInfoText:[NSString stringWithFormat:@"carrier name: %@", provider.carrierName]];

    // 国コードISO
    NSString *isoCountry = provider.isoCountryCode;
    if (isoCountry.length > 0) {
        [self addInfoText:[NSString stringWithFormat:@"ISO Country code: %@", isoCountry]];
    }

    // 国コードPreferredLanguage
    NSString *countryCode = [[NSLocale preferredLanguages] objectAtIndex:0];
    if (countryCode.length > 0) {
        [self addInfoText:[NSString stringWithFormat:@"PreferredLanguage: %@", countryCode]];
    }

    // 言語コード
    NSString *localeCode = [[NSLocale currentLocale] objectForKey:NSLocaleIdentifier];
    if (localeCode.length > 0) {
        [self addInfoText:[NSString stringWithFormat:@"code: %@", localeCode]];
    }

    // BundleID
    NSString *bundleId = [[NSBundle mainBundle] bundleIdentifier];
    [self addInfoText:[NSString stringWithFormat:@"BundleID: %@", bundleId]];

    // サポートOSバージョン
    [self addInfoText:[NSString stringWithFormat:@"サポート対象OS: %@",
                       b2s(VAMP.isSupported)]];
    [self addInfoText:[NSString stringWithFormat:@"ChildDirected: %@",
                       [self childDirectedString]]];

}

- (void) viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];

    self.adInfoView.frame = self.view.frame;
}

- (void) addInfoText:(NSString *)text {
    self.adInfoView.text = [self.adInfoView.text stringByAppendingFormat:@"\n%@", text];
}

- (NSString *) adNetworkVersion:(NSString *)adapterClass {
    Class cls = NSClassFromString(adapterClass);

    if (!cls) {
        return @"-";
    }
    SEL sel = NSSelectorFromString(@"adNetworkVersion");
    IMP method = [cls instanceMethodForSelector:sel];
    AdNetworkVersionMethod func = (void *) method;
    id obj = [cls new];
    return func(obj, sel);
}

- (NSString *) adapterVersion:(NSString *)adapterClass {
    Class cls = NSClassFromString(adapterClass);

    if (!cls) {
        return @"-";
    }
    SEL sel = NSSelectorFromString(@"adapterVersion");
    IMP method = [cls instanceMethodForSelector:sel];
    AdapterVersionMethod func = (void *) method;
    id obj = [cls new];
    NSString *ver = func(obj, sel);

    // アダプタバージョンは以下の形式のため、
    // @(#)PROGRAM:VAMPAdMobAdapter  PROJECT:VAMPAdMobAdapter-9.3.0.0\n
    // 最後のバージョン番号だけを抽出する
    NSArray<NSString *> *comps = [ver componentsSeparatedByString:@"-"];
    return [comps.lastObject
            stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet] ?: @"";
}

- (NSString *) childDirectedString {
    VAMPChildDirected childDirected = VAMPPrivacySettings.childDirected;

    switch (childDirected) {
        case kVAMPChildDirectedTrue:
            return @"true";
        case kVAMPChildDirectedFalse:
            return @"false";
        case kVAMPChildDirectedUnknown:
            return @"unknown";
        default:
            return [NSString stringWithFormat:@"%ld", (long) childDirected];
    }
}
@end

File: VAMPSwiftSample/VAMPSwiftSample/InfoViewController.swift
//
//  InfoViewController.swift
//  VAMPSwiftSample
//
//  Created by Supership Inc. on 2023/01/28.
//  Copyright © 2023 Supership Inc. All rights reserved.
//

import CoreTelephony
import Foundation
import UIKit

import VAMP

class InfoViewController: UIViewController {
    var adInfoView: UITextView

    class func instantiate() -> InfoViewController {
        InfoViewController()
    }

    init() {
        adInfoView = UITextView()
        adInfoView.textContainerInset = UIEdgeInsets(top: 0, left: 16,
                                                     bottom: 0, right: 16)
        adInfoView.isEditable = false
        adInfoView.isSelectable = false
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        title = "Info"

        view.addSubview(adInfoView)
        adInfoView.text = ""

        // VAMP SDK Version
        addInfoText(text: String(format: "VAMPSDK: %@", VAMPSDKVersion))

        // アドネットワークバージョン
        addInfoText(text: String(format: "AdMobSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPAdMobSDKAdapter")))
        addInfoText(text: String(format: "maioSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPMaioSDKAdapter")))
        addInfoText(text: String(format: "UnityAdsSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPUnityAdsSDKAdapter")))
        addInfoText(text: String(format: "PangleSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPPangleSDKAdapter")))
        addInfoText(text: String(format: "LINEAdsSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPLINEAdsSDKAdapter")))
        addInfoText(text: String(format: "IronSourceSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPIronSourceSDKAdapter")))
        addInfoText(text: String(format: "AppLovinSDK: %@",
                                 adNetworkVersion(adapterClass: "VAMPAppLovinSDKAdapter")))

        addInfoText(text: "\n")

        // アダプタバージョン
        addInfoText(text: String(format: "AdMob Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPAdMobSDKAdapter")))
        addInfoText(text: String(format: "maio Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPMaioSDKAdapter")))
        addInfoText(text: String(format: "UnityAds Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPUnityAdsSDKAdapter")))
        addInfoText(text: String(format: "Pangle Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPPangleSDKAdapter")))
        addInfoText(text: String(format: "LINEAds Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPLINEAdsSDKAdapter")))
        addInfoText(text: String(format: "IronSource Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPIronSourceSDKAdapter")))
        addInfoText(text: String(format: "AppLovin Adapter: %@",
                                 adapterVersion(adapterClass: "VAMPAppLovinSDKAdapter")))

        addInfoText(text: "\n")

        addInfoText(text: String(format: "useHyperID: %@",
                                 VAMP.useHyperID().description))

        addInfoText(text: "\n")

        // デバイス名
        addInfoText(text: String(format: "device name: %@",
                                 UIDevice.current.name))

        // OS名
        addInfoText(text: String(format: "system name: %@",
                                 UIDevice.current.systemName))

        // OSバージョン
        addInfoText(text: String(format: "system version: %@",
                                 UIDevice.current.systemVersion))

        // OSモデル
        addInfoText(text: String(format: "OS model: %@",
                                 UIDevice.current.model))

        // キャリア情報
        let networkInfo = CTTelephonyNetworkInfo()
        let provider = networkInfo.subscriberCellularProvider
        addInfoText(text: String(format: "carrier name: %@",
                                 provider?.carrierName ?? "-"))

        // 国コードISO
        addInfoText(text: String(format: "ISO Country code: %@",
                                 provider?.isoCountryCode ?? "-"))

        // 国コードPreferredLanguage
        addInfoText(text: String(format: "PreferredLanguage: %@",
                                 NSLocale.preferredLanguages.first ?? "-"))

        // 言語コード
        addInfoText(text: String(format: "code: %@",
                                 Locale.current.languageCode ?? "-"))

        // BundleID
        addInfoText(text: String(format: "BundleID: %@",
                                 Bundle.main.bundleIdentifier ?? "-"))

        // サポートOSバージョン
        addInfoText(text: String(format: "サポート対象OS: %@",
                                 VAMP.isSupported().description))
        addInfoText(text: String(format: "ChildDirected: %@",
                                 VAMPPrivacySettings.childDirected().toString))
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()

        adInfoView.frame = view.frame
    }

    func addInfoText(text: String) {
        adInfoView.text = adInfoView.text.appendingFormat("\n%@", text)
    }

    func adNetworkVersion(adapterClass: String) -> String {
        let cls = NSClassFromString(adapterClass) as? NSObject.Type
        if cls == nil {
            return "-"
        }

        let obj = cls?.init() as? VAMPAdapter
        return obj?.adNetworkVersion() ?? "-"
    }

    func adapterVersion(adapterClass: String) -> String {
        let cls = NSClassFromString(adapterClass) as? NSObject.Type
        if cls == nil {
            return "-"
        }

        let obj = cls?.init() as? VAMPAdapter
        let version = obj?.adapterVersion() ?? "-"

        // アダプタバージョンは以下の形式のため、
        // @(#)PROGRAM:VAMPAdMobAdapter  PROJECT:VAMPAdMobAdapter-9.3.0.0\n
        // 最後のバージョン番号だけを抽出する
        let comps = version.components(separatedBy: "-")
        return comps.last?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? "-"
    }
}

extension VAMPChildDirected {
    var toString: String {
        switch self {
            case .true:
                return "true"
            case .false:
                return "false"
            case .unknown:
                return "unknown"
        }
    }
}
